_posts/2019-03-11-dask-nep18.md

```diff
@@ -0,0 +1,521 @@
+we chose the
+[`lbfgs`](https://docs.scipy.org/doc/scipy/reference/optimize.minimize-lbfgsb.html)
+solver, that converges reasonably fast.
+
+We can also have a look at the timing results for fitting, followed by those
+for predicting with Dask-GLM:
+
+```
+# Fitting
+CPU times: user 9.66 ms, sys: 116 µs, total: 9.78 ms
+Wall time: 8.94 ms
+
+# Predicting
+CPU times: user 130 µs, sys: 327 µs, total: 457 µs
+Wall time: 1.06 ms
+```
+
+If instead we want to use CuPy to compute, we have to change only 3 lines,
+importing `cupy` instead of `numpy`, and the two lines where we create the
+random arrays, replacing them to `cupy.random` insted of `np.random`. The
+block should then look like this:
+
+```python
+import cupy
+from dask_glm.estimators import LinearRegression
+import matplotlib.pyplot as plt
+
+N = 1000
+
+# x from 0 to N
+x = N * cupy.random.random((40000, 1))
+
+# y = a*x + b with noise
+y = 0.5 * x + 1.0 + cupy.random.normal(size=x.shape)
+
+# create a linear regression model
+est = LinearRegression(solver='lbfgs')
+```
+
+And the timing results we obtain in this scenario are:
+
+```
+# Fitting
+CPU times: user 151 ms, sys: 40.7 ms, total: 191 ms
+Wall time: 190 ms
+
+# Predicting
+CPU times: user 1.91 ms, sys: 778 µs, total: 2.69 ms
+Wall time: 1.37 ms
+```
+
+For the simple example chosen for this post, scikit-learn outperforms Dask-GLM
+using both NumPy and CuPy arrays. There may exist several reasons that
+contribute to this, and while we didn't dive deep into understanding the exact
+reasons and their extent, we could cite some likely possibilities:
+
+*   scikit-learn may be using solvers that converge faster;
+*   Dask-GLM is entirely built on top of Dask, while scikit-learn may be
+    heavily optimized internally;
+*   Too many synchronization steps and data transfer could occur for small
+    datasets with CuPy.
+
+### Performance for different Dask-GLM solvers
+
+To verify how Dask-GLM with Dask arrays would compare with CuPy arrays, we also
+did some linear regression benchmarking of Dask-GLM solvers. The results below
+were obtained from a training dataset with 1000 features of 100 dimensions,
+and 1000 test features.
+
+```
+Solver admm with Dask took 22444.307 ms to fit and 41.299 ms to predict
+Solver admm with CuPy took 263.742 ms to fit and 1.146 ms to predict
+
+Solver lbfgs with Dask took 1561.047 ms to fit and 41.063 ms to predict
+Solver lbfgs with CuPy took 13.935 ms to fit and 0.260 ms to predict
+
+Solver newton with Dask took 785.150 ms to fit and 42.822 ms to predict
+Solver newton with CuPy took 22.480 ms to fit and 0.248 ms to predict
+
+Solver proximal_grad with Dask took 1902.620 ms to fit and 45.349 ms to predict
+Solver proximal_grad with CuPy took 11.700 ms to fit and 0.257 ms to predict
+
+Solver gradient_descent with Dask took 2483.603 ms to fit and 42.605 ms to predict
+Solver gradient_descent with CuPy took 14.580 ms to fit and 0.257 ms to predict
```

Comment:
When you say "Dask" above you mean Dask+NumPy, correct?  Would it be hard to include straight NumPy numbers here as well?

Reply:
You may also want to format this as a table.  I usually use HTML for this rather than markdown.  

https://github.com/dask/dask-blog/blob/gh-pages/_posts/2019-01-03-dask-array-gpus-first-steps.md#results

Reply:
> Would it be hard to include straight NumPy numbers here as well?

I suppose not, I will check that.

> You may also want to format this as a table. I usually use HTML for this rather than markdown.

Great idea, thanks for the suggestion!

Reply:
Well, that's very unexpected. Using NumPy arrays is faster than CuPy, and _much_ faster than Dask:

```
Solver admm took 120.516 ms to fit and 0.492 ms to predict
Solver lbfgs took 13.858 ms to fit and 0.438 ms to predict
Solver newton took 52.420 ms to fit and 0.235 ms to predict
Solver proximal_grad took 4.047 ms to fit and 0.216 ms to predict
Solver gradient_descent took 6.207 ms to fit and 0.218 ms to predict
```

I could understand NumPy and CuPy being more or less close, but not Dask being absurdly slower than both. The code is basically this (plus imports and timing computation):

```python
X, y = make_classification()
lr.fit(X, y)
lr.predict(X)
```

Reply:
After a bit of profiling, this is what I found out for Dask-GLM with Dask array:

```
    14339    0.139    0.000    0.814    0.000 /home/pentschev/.local/lib/python3.5/site-packages/dask/local.py:430(fire_task)
    44898   19.945    0.000   19.945    0.000 {method 'acquire' of '_thread.lock' objects}
     4055    0.042    0.000   19.992    0.005 /usr/lib/python3.5/threading.py:261(wait)
    14339    0.107    0.000   20.234    0.001 /usr/lib/python3.5/queue.py:147(get)
    14339    0.018    0.000   20.253    0.001 /home/pentschev/.local/lib/python3.5/site-packages/dask/local.py:140(queue_get)
      122    0.117    0.001   22.327    0.183 /home/pentschev/.local/lib/python3.5/site-packages/dask/local.py:345(get_async)
      122    0.013    0.000   22.346    0.183 /home/pentschev/.local/lib/python3.5/site-packages/dask/threaded.py:33(get)
      122    0.004    0.000   22.733    0.186 /home/pentschev/.local/lib/python3.5/site-packages/dask/base.py:345(compute)
        1    0.020    0.020   23.224   23.224 /home/pentschev/.local/lib/python3.5/site-packages/dask_glm/algorithms.py:200(admm)
        1    0.000    0.000   23.267   23.267 /home/pentschev/.local/lib/python3.5/site-packages/dask_glm/utils.py:13(normalize_inputs)
        1    0.000    0.000   23.268   23.268 /home/pentschev/.local/lib/python3.5/site-packages/dask_glm/estimators.py:65(fit)
```

A big portion of the time seems to be spent on waiting for thread lock. Also, looking at the callers, we see 100 `compute()` calls departing from `admm()`, which means it's _not_ converging and stopping only at `max_iter` as @cicdw suggested:

```
/home/pentschev/.local/lib/python3.5/site-packages/dask/base.py:345(compute)                               <-     100    0.004   19.637  /home/pentschev/.local/lib/python3.5/site-packages/dask_glm/algorithms.py:197(admm)
```

Running with NumPy, the algorithm converges, showing only 7 `compute()` calls:

```
/home/pentschev/.local/lib/python3.5/site-packages/dask/base.py:345(compute)                          <-       7    0.000    0.120  /home/pentschev/.local/lib/python3.5/site-packages/dask_glm/algorithms.py:197(admm)
```

I'm running Dask 1.1.4 and Dask-GLM master branch, to ensure that my local changes aren't introduce any bugs. However, if I run my Dask-GLM branch and use CuPy as a backend, it also converges in 7 iterations.

To me this seems to suggest that we have one of those very well-hidden and difficult to track bugs in Dask. Before I spent hours with this, any suggestions what could we look for?

Reply:
All multi-threaded code profiles as "waiting for thread lock" in the main thread.  It's hard to profile multi-threaded code in Python.  If you're going to profile I recommend using `.compute(scheduler="single-threaded")`

Reply:
I know it's harder to profile multithreaded code. But profiling's not the point from my last comment, it only helped me see that when we use Dask arrays, the algorithms don't converge, and that's definitely the main issue with it taking so much longer.

I'm now asking for ideas on obvious things I could look for, profiling is irrelevant if we really have a bug that needs to be fixed, and that's why I was pointing out that NumPy and CuPy both converge, but Dask does not.

Minimal code to reproduce:

```python
from dask_glm.estimators import LogisticRegression
from dask_glm.datasets import make_classification

X, y = make_classification()

lr = LogisticRegression(solver='admm')
lr.fit(X, y)
```

Reply:
OK, we should fix that.  To be honest I'm not sure how at the moment.  

For the purposes of this blogpost perhaps it would be better to compare just numpy vs cupy, and leave dask out of it?

Reply:
And to run the code above on a NumPy array (that converges), it suffices to change the last line to:

```python
lr.fit(X.compute(), y.compute())
```

Reply:
cc @cicdw if you have suggestions

Reply:
> OK, we should fix that. To be honest I'm not sure how at the moment.

Yes, that's alright, I was mainly interested if anybody knew of something obvious.

> For the purposes of this blogpost perhaps it would be better to compare just numpy vs cupy, and leave dask out of it?

Agreed. I will do that and mention we currently have a bug in Dask, that's why we're leaving it out.

Reply:
> mention we currently have a bug in Dask

Just to be clear,my guess is that it's more likely a bug in dask-glm than dask itself.

Reply:
It could be too, but my first suspicion was Dask because both NumPy and CuPy converge, so Dask becomes the obvious suspect, which doesn't make it necessarily guilty.

Reply:
Also, Dask-GLM estimators were deprecated in https://github.com/dask/dask-glm/pull/66. Maybe they indeed have a bug which may or may not have been fixed in Dask-ML.

Reply:
I don't actually think this is a bug anywhere; ADMM is implemented to work with potentially distributed data, and works by solving a sequence of subproblems on the different pieces of the data (and is only truly useful for _approximate_ large scale solutions).

I'm guessing that the partioning of your data when using dask.array is just not a good partioning.  (Relevant line of code: https://github.com/dask/dask-glm/blob/master/dask_glm/algorithms.py#L241)

Reply:
@cicdw you may be right, but then we would have a "bug" in the `make_classification()` or in the example provided in https://dask-glm.readthedocs.io/en/latest/estimators.html.

Reply:
I also didn't check every single solver, but they all take considerably longer with Dask than NumPy, so possibly all of them aren't converging as fast as with NumPy arrays.